File: frontend/js/views/tax_aware_rebalance.js
// Tax-Aware Rebalance — trade toward target weights while choosing sell
// lots to minimize realized gain (or harvest losses). Enter one row per
// tax lot; rows are grouped by symbol (price + target taken from the
// first lot of each symbol), then planned server-side via
// /calc/tax-aware-rebalance.

import { api } from '../api.js';
import { esc } from '../util.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';

const LS_KEY = 'tv-tax-rebalance-lots-v1';

function load() {
    try {
        const raw = localStorage.getItem(LS_KEY);
        return raw ? JSON.parse(raw) : [];
    } catch { return []; }
}
function save(lots) {
    try { localStorage.setItem(LS_KEY, JSON.stringify(lots)); } catch { /* private mode */ }
}

const state = { lots: load(), strategy: 'hifo', tax_rate: 25, band: 5 };

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });

export async function renderTaxAwareRebalance(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.tax_rebal.h1.title">// TAX-AWARE REBALANCE</span></h1>
        <p class="muted small" data-i18n="view.tax_rebal.hint.intro">
            Rebalancing to a target can realize a large capital gain it never tells you
            about. Enter one row per tax lot; the planner trades toward your targets but
            picks sell lots to minimize the gain (HIFO) or to harvest losses, and reports
            the gain and estimated tax the rebalance triggers. Holdings within the no-trade
            band are left alone — churning them realizes tax for a trivial correction.
        </p>
        <div class="chart-panel">
            <h2 data-i18n="view.tax_rebal.h2.add_lot">Add tax lot</h2>
            <form id="tr-form" class="inline-form">
                <input name="symbol" placeholder="symbol" data-i18n-placeholder="common.placeholder.symbol" required style="text-transform:uppercase">
                <input name="qty" type="number" step="0.0001" min="0" placeholder="shares" data-i18n-placeholder="view.tax_rebal.ph.shares" required>
                <input name="cost" type="number" step="0.01" min="0" placeholder="cost/share" data-i18n-placeholder="view.tax_rebal.ph.cost" required>
                <input name="price" type="number" step="0.01" min="0" placeholder="price now" data-i18n-placeholder="view.tax_rebal.ph.price" required>
                <input name="target" type="number" step="0.1" min="0" max="100" placeholder="target %" data-i18n-placeholder="view.tax_rebal.ph.target" required>
                <button class="primary" type="submit" data-i18n="view.tax_rebal.btn.add">Add lot</button>
            </form>
        </div>
        <div class="chart-panel">
            <h2 data-i18n="view.tax_rebal.h2.settings">Plan settings</h2>
            <form id="tr-settings" class="inline-form">
                <label><span data-i18n="view.tax_rebal.label.strategy">Sell-lot strategy</span>
                    <select name="strategy">
                        <option value="hifo" data-i18n="view.tax_rebal.opt.hifo">HIFO — minimize gain</option>
                        <option value="max_loss_harvest" data-i18n="view.tax_rebal.opt.harvest">Harvest losses only</option>
                        <option value="lifoust" data-i18n="view.tax_rebal.opt.lowcost">Lowest-cost — maximize gain</option>
                    </select>
                </label>
                <label><span data-i18n="view.tax_rebal.label.tax_rate">Tax rate %</span>
                    <input type="number" step="0.1" min="0" max="100" name="tax_rate" value="${state.tax_rate}"></label>
                <label><span data-i18n="view.tax_rebal.label.band">No-trade band %</span>
                    <input type="number" step="0.1" min="0" max="50" name="band" value="${state.band}"></label>
                <button class="primary" type="submit" data-i18n="view.tax_rebal.btn.plan">Plan rebalance</button>
            </form>
        </div>
        <div id="tr-result"></div>
        <div id="tr-table" class="chart-panel"></div>
    `;
    applyUiI18n(mount);

    mount.querySelector('#tr-form').addEventListener('submit', (e) => {
        e.preventDefault();
        const fd = new FormData(e.target);
        state.lots.push({
            id: crypto.randomUUID ? crypto.randomUUID() : String(Date.now()) + Math.random(),
            symbol: (fd.get('symbol') || '').trim().toUpperCase(),
            qty: Number(fd.get('qty')),
            cost: Number(fd.get('cost')),
            price: Number(fd.get('price')),
            target: Number(fd.get('target')),
        });
        save(state.lots);
        e.target.reset();
        renderTable(mount);
    });

    mount.querySelector('#tr-settings').addEventListener('submit', async (e) => {
        e.preventDefault();
        const fd = new FormData(e.target);
        state.strategy = fd.get('strategy');
        state.tax_rate = Number(fd.get('tax_rate')) || 0;
        state.band = Number(fd.get('band')) || 0;
        await runPlan(mount, tok);
    });

    renderTable(mount);
}

// Group flat lot rows into the holdings[] the API expects: price and
// target come from the first lot seen for each symbol.
function groupHoldings() {
    const bySym = new Map();
    for (const l of state.lots) {
        if (!bySym.has(l.symbol)) {
            bySym.set(l.symbol, {
                symbol: l.symbol,
                price: l.price,
                target_weight: l.target / 100,
                lots: [],
            });
        }
        bySym.get(l.symbol).lots.push({
            lot_id: l.id,
            qty_open: l.qty,
            cost_per_share: l.cost,
        });
    }
    return [...bySym.values()];
}

async function runPlan(mount, tok) {
    const holdings = groupHoldings();
    if (!holdings.length) {
        showToast(t('view.tax_rebal.toast.no_lots'), { level: 'warning' });
        return;
    }
    try {
        const plan = await api.calcTaxAwareRebalance({
            holdings,
            strategy: state.strategy,
            tax_rate: state.tax_rate / 100,
            band: state.band / 100,
        });
        if (!viewIsCurrent(tok)) return;
        renderResult(mount, plan);
    } catch (err) {
        showToast(err.message || t('view.tax_rebal.toast.error'), { level: 'error' });
    }
}

function renderResult(mount, plan) {
    const el = mount.querySelector('#tr-result');
    const gainCls = Number(plan.total_realized_gain) >= 0 ? 'neg' : 'pos'; // a gain is a tax cost; a loss is a benefit
    const rows = plan.actions.map((a) => {
        const tv = Number(a.trade_value);
        const verb = tv > 0 ? t('view.tax_rebal.act.buy') : tv < 0 ? t('view.tax_rebal.act.sell') : t('view.tax_rebal.act.hold');
        const cls = tv > 0 ? 'pos' : tv < 0 ? 'neg' : 'muted';
        const g = Number(a.realized_gain);
        return `<tr>
            <td>${esc(a.symbol)}</td>
            <td>${money(a.current_value)}</td>
            <td>${money(a.target_value)}</td>
            <td class="${cls}">${verb} ${tv === 0 ? '' : money(Math.abs(tv))}</td>
            <td class="${g > 0 ? 'neg' : g < 0 ? 'pos' : 'muted'}">${g === 0 ? '—' : money(g)}</td>
        </tr>`;
    }).join('');
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.tax_rebal.h2.plan">Rebalance plan</h2>
            <div class="cards">
                <div class="card"><div class="label" data-i18n="view.tax_rebal.card.total">Portfolio value</div>
                    <div class="value">${money(plan.total_value)}</div></div>
                <div class="card"><div class="label" data-i18n="view.tax_rebal.card.gain">Realized gain</div>
                    <div class="value ${gainCls}">${money(plan.total_realized_gain)}</div></div>
                <div class="card"><div class="label" data-i18n="view.tax_rebal.card.tax">Estimated tax</div>
                    <div class="value neg">${money(plan.estimated_tax)}</div></div>
            </div>
            <table class="trades">
                <thead><tr>
                    <th data-i18n="view.tax_rebal.th.symbol">Symbol</th>
                    <th data-i18n="view.tax_rebal.th.current">Current</th>
                    <th data-i18n="view.tax_rebal.th.target">Target</th>
                    <th data-i18n="view.tax_rebal.th.trade">Trade</th>
                    <th data-i18n="view.tax_rebal.th.gain">Realized gain</th>
                </tr></thead>
                <tbody>${rows}</tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

function renderTable(mount) {
    const el = mount.querySelector('#tr-table');
    if (!el) return;
    if (!state.lots.length) {
        el.innerHTML = `<h2 data-i18n="view.tax_rebal.h2.lots">Tax lots</h2>
            <p class="muted" data-i18n="view.tax_rebal.empty">No lots yet.</p>`;
        applyUiI18n(el);
        return;
    }
    el.innerHTML = `
        <h2 data-i18n="view.tax_rebal.h2.lots">Tax lots</h2>
        <table class="trades">
            <thead><tr>
                <th data-i18n="view.tax_rebal.th.symbol">Symbol</th>
                <th data-i18n="view.tax_rebal.th.shares">Shares</th>
                <th data-i18n="view.tax_rebal.th.cost">Cost/sh</th>
                <th data-i18n="view.tax_rebal.th.price">Price</th>
                <th data-i18n="view.tax_rebal.th.target">Target %</th>
                <th data-i18n="view.tax_rebal.th.actions">Actions</th>
            </tr></thead>
            <tbody>${state.lots.map((l) => `<tr>
                <td>${esc(l.symbol)}</td>
                <td>${l.qty}</td>
                <td>$${l.cost}</td>
                <td>$${l.price}</td>
                <td>${l.target}%</td>
                <td><button class="link neg" data-del="${esc(l.id)}" data-i18n="view.tax_rebal.btn.delete">delete</button></td>
            </tr>`).join('')}</tbody>
        </table>
    `;
    applyUiI18n(el);
    el.querySelectorAll('[data-del]').forEach((btn) => {
        btn.addEventListener('click', () => {
            state.lots = state.lots.filter((l) => l.id !== btn.dataset.del);
            save(state.lots);
            renderTable(mount);
        });
    });
}

File: crates/traderview-core/src/tax_aware_rebalance.rs
//! Tax-aware rebalance — trade toward target weights while minimizing
//! the tax the rebalance realizes.
//!
//! The plain rebalancer only cares about weights; selling to hit a target
//! can realize a large capital gain it never reports. This plans the same
//! drift correction but, on the SELL side, chooses tax lots by strategy
//! (HIFO to minimize gain, MaxLossHarvest to realize losses) via the
//! shared [`crate::tax_lot_optimizer`], reports the gain each sale
//! triggers, and leaves holdings whose drift is within a no-trade band
//! untouched — trading them would realize tax for a trivial correction.

use crate::tax_lot_optimizer::{close, ClosingEntry, CostLot, LotStrategy};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct Holding {
    pub symbol: String,
    /// Current price per share.
    pub price: Decimal,
    /// Target portfolio weight as a fraction in [0, 1].
    pub target_weight: Decimal,
    /// Open tax lots for this holding.
    pub lots: Vec<CostLot>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Action {
    pub symbol: String,
    pub current_value: Decimal,
    pub target_value: Decimal,
    /// Signed traded value: positive = buy, negative = sell, zero = held
    /// (within band). For sells it is the value actually sold, which can
    /// be less than the drift when MaxLossHarvest runs out of loss lots.
    pub trade_value: Decimal,
    /// Capital gain the sale realizes (zero for buys and holds; negative
    /// when harvesting losses).
    pub realized_gain: Decimal,
    pub closed_lots: Vec<ClosingEntry>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Plan {
    pub total_value: Decimal,
    pub actions: Vec<Action>,
    pub total_realized_gain: Decimal,
    /// `tax_rate` applied to the NET realized gain. A net loss yields
    /// zero tax — it's a harvested benefit, not a liability.
    pub estimated_tax: Decimal,
}

/// Plan a tax-aware rebalance. `band` is the no-trade threshold as a
/// fraction of total value: a holding within `band` of its target is
/// left alone. `strategy` picks the sell lots; `tax_rate` (fraction)
/// estimates the tax on the net realized gain.
pub fn plan(
    holdings: &[Holding],
    strategy: LotStrategy,
    tax_rate: Decimal,
    band: Decimal,
) -> Plan {
    let total_value: Decimal = holdings
        .iter()
        .map(|h| shares(h) * h.price)
        .sum();

    let mut actions = Vec::with_capacity(holdings.len());
    let mut total_realized_gain = Decimal::ZERO;

    for h in holdings {
        let current_value = shares(h) * h.price;
        let target_value = h.target_weight * total_value;
        let drift = target_value - current_value;

        let within_band =
            total_value > Decimal::ZERO && (drift.abs() / total_value) < band;

        let (trade_value, realized_gain, closed_lots) = if within_band || drift.is_zero() {
            (Decimal::ZERO, Decimal::ZERO, Vec::new())
        } else if drift > Decimal::ZERO {
            // Underweight → buy. No tax event.
            (drift, Decimal::ZERO, Vec::new())
        } else if h.price > Decimal::ZERO {
            // Overweight → sell `drift` worth, choosing lots by strategy.
            let qty_to_sell = (-drift) / h.price;
            let report = close(&h.lots, qty_to_sell, h.price, strategy);
            let sold_qty = qty_to_sell - report.qty_remaining_to_close;
            (-(sold_qty * h.price), report.total_realized_gain, report.entries)
        } else {
            (Decimal::ZERO, Decimal::ZERO, Vec::new())
        };

        total_realized_gain += realized_gain;
        actions.push(Action {
            symbol: h.symbol.clone(),
            current_value,
            target_value,
            trade_value,
            realized_gain,
            closed_lots,
        });
    }

    let estimated_tax = total_realized_gain.max(Decimal::ZERO) * tax_rate;
    Plan { total_value, actions, total_realized_gain, estimated_tax }
}

fn shares(h: &Holding) -> Decimal {
    h.lots.iter().map(|l| l.qty_open).sum()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn d(v: i64) -> Decimal {
        Decimal::from(v)
    }
    fn dq(s: &str) -> Decimal {
        s.parse().unwrap()
    }
    fn lot(id: &str, qty: i64, cost: i64) -> CostLot {
        CostLot { lot_id: id.into(), qty_open: d(qty), cost_per_share: d(cost) }
    }

    // 100 sh AAPL @ $100 (cost $50) = $10k; 100 sh MSFT @ $100 (cost $90)
    // = $10k. Total $20k, currently 50/50.
    fn two_holdings(aapl_target: &str, msft_target: &str) -> Vec<Holding> {
        vec![
            Holding {
                symbol: "AAPL".into(),
                price: d(100),
                target_weight: dq(aapl_target),
                lots: vec![lot("a1", 100, 50)],
            },
            Holding {
                symbol: "MSFT".into(),
                price: d(100),
                target_weight: dq(msft_target),
                lots: vec![lot("m1", 100, 90)],
            },
        ]
    }

    #[test]
    fn already_balanced_no_trades() {
        let p = plan(&two_holdings("0.5", "0.5"), LotStrategy::Hifo, dq("0.2"), Decimal::ZERO);
        assert_eq!(p.total_value, d(20_000));
        for a in &p.actions {
            assert_eq!(a.trade_value, Decimal::ZERO);
        }
        assert_eq!(p.total_realized_gain, Decimal::ZERO);
        assert_eq!(p.estimated_tax, Decimal::ZERO);
    }

    #[test]
    fn shift_to_70_30_sells_overweight_and_realizes_gain() {
        // Target AAPL 30% ($6k) / MSFT 70% ($14k). AAPL is overweight by
        // $4k → sell 40 sh @ $100, cost $50 → gain $50×40 = $2,000.
        let p = plan(&two_holdings("0.3", "0.7"), LotStrategy::Hifo, dq("0.2"), Decimal::ZERO);
        let aapl = p.actions.iter().find(|a| a.symbol == "AAPL").unwrap();
        let msft = p.actions.iter().find(|a| a.symbol == "MSFT").unwrap();
        assert_eq!(aapl.trade_value, d(-4_000)); // sold $4k
        assert_eq!(aapl.realized_gain, d(2_000)); // (100-50)*40
        assert_eq!(msft.trade_value, d(4_000)); // bought $4k, no gain
        assert_eq!(msft.realized_gain, Decimal::ZERO);
        assert_eq!(p.total_realized_gain, d(2_000));
    }

    #[test]
    fn tax_rate_applies_to_net_gain() {
        let p = plan(&two_holdings("0.3", "0.7"), LotStrategy::Hifo, dq("0.25"), Decimal::ZERO);
        assert_eq!(p.estimated_tax, d(500)); // 2000 * 0.25
    }

    #[test]
    fn loss_harvest_realizes_loss_and_zero_tax() {
        // MSFT cost $90, price $100 → no loss there. Build a losing lot.
        let holdings = vec![
            Holding {
                symbol: "TSLA".into(),
                price: d(100),
                target_weight: dq("0.5"),
                lots: vec![lot("t1", 100, 130)], // $130 cost, now $100 → loss
            },
            Holding {
                symbol: "NVDA".into(),
                price: d(100),
                target_weight: dq("0.5"),
                lots: vec![lot("n1", 100, 40)],
            },
        ];
        // Total $20k, 50/50 already — force a sell by shifting to 30/70.
        let mut h = holdings;
        h[0].target_weight = dq("0.3");
        h[1].target_weight = dq("0.7");
        let p = plan(&h, LotStrategy::MaxLossHarvest, dq("0.25"), Decimal::ZERO);
        let tsla = p.actions.iter().find(|a| a.symbol == "TSLA").unwrap();
        // Sells 40 sh of the $130 lot → loss (100-130)*40 = -1200.
        assert_eq!(tsla.realized_gain, d(-1_200));
        assert!(p.total_realized_gain < Decimal::ZERO);
        assert_eq!(p.estimated_tax, Decimal::ZERO); // a net loss is not taxed
    }

    #[test]
    fn no_trade_band_skips_small_drift() {
        // 52/48 vs a 50/50 target is 2% drift; a 5% band leaves it alone.
        let p = plan(&two_holdings("0.52", "0.48"), LotStrategy::Hifo, dq("0.25"), dq("0.05"));
        for a in &p.actions {
            assert_eq!(a.trade_value, Decimal::ZERO, "{} should be within band", a.symbol);
        }
    }
}
